File: src/spektrafilm/runtime/pipeline.py
from __future__ import annotations

import copy
from time import perf_counter

import numpy as np

from spektrafilm.runtime.services import (
    EnlargerService,
    ResizingService,
    SpectralLUTService,
    ColorReferenceService,
)
from spektrafilm.runtime.stages import FilmingStage, PrintingStage, ScanningStage
from spektrafilm.utils.timings import format_timings



class SimulationPipeline:
    """Thin runtime orchestrator that composes stage objects."""

    def __init__(self, params, update_params=False):
        self._params = copy.deepcopy(params)

        self.camera = self._params.camera
        self.film = self._params.film
        self.film_render = self._params.film_render
        self.enlarger = self._params.enlarger
        self.print = self._params.print
        self.print_render = self._params.print_render
        self.scanner = self._params.scanner
        self.io = self._params.io
        self.debug = self._params.debug
        self.settings = self._params.settings

        self.timings = {}
        self._last_elapsed_time = None

        self._resize_service = ResizingService(self.io, self.camera.film_format_mm)
        if not update_params:
            self._lut_service = SpectralLUTService(self.settings.lut_resolution)
        self._enlarger_service = EnlargerService(self.enlarger)
        self._color_reference_service = ColorReferenceService(self.film, self.film_render,
                                                              self.print, self.print_render,
                                                              self.scanner.black_correction, self.scanner.white_correction,
                                                              self.scanner.black_level, self.scanner.white_level,
                                                              self.io)

        
        self._filming_stage = FilmingStage(
            self.film,
            self.film_render,
            self.camera,
            self.io,
            self.settings,
            self._lut_service,
            self._resize_service, # to get pixel size um for blurs
            self._enlarger_service, # to compute and save density spectral midgray to balance print
            self._color_reference_service,
        )
        self._printing_stage = PrintingStage(
            self.film,
            self.film_render,
            self.print,
            self.print_render,
            self.enlarger,
            self.settings,
            self._lut_service,
            self._enlarger_service,
            self._resize_service, # to get pixel size um for diffusion filter
            self._color_reference_service,
        )
        self._scanning_stage = ScanningStage(
            self.film,
            self.film_render,
            self.print,
            self.print_render,
            self.scanner,
            self.io,
            self.settings,
            self._lut_service,
            self._color_reference_service,
        )
        
        # timing communication
        self._filming_stage.timings = self.timings
        self._printing_stage.timings = self.timings
        self._scanning_stage.timings = self.timings
        self._lut_service.timings = self.timings

    def process(self, image):
        """Process an image through the simulation pipeline."""
        self.timings.clear()
        start = perf_counter()
        try:
            if self.debug.debug_mode == 'off':
                image = self._pipeline(image)
            else:
                image = self._pipeline_debug(image)
            return image
        finally:
            self._last_elapsed_time = perf_counter() - start

    def get_timings(self):
        return self.timings

    def get_total_elapsed_time(self):
        return self._last_elapsed_time

    def format_timings(self):
        return format_timings(
            self.get_timings(),
            total_elapsed_time=self.get_total_elapsed_time(),
        )

    def print_timings(self):
        print(self.format_timings())
    
    def update(self, params):
        """Update params and re-initialize stages that depend on them."""
        self.__init__(params, update_params=True)
        
    def soft_update(self,
                    exposure_compensation_ev=None,
                    print_exposure=None,
                    c_filter_neutral=None,
                    m_filter_neutral=None,
                    y_filter_neutral=None,
                    film_density_curves=None,
                    print_density_curves=None,):
        invalidates_print_balance_reference = False
        if exposure_compensation_ev is not None:
            self.camera.exposure_compensation_ev = exposure_compensation_ev
            invalidates_print_balance_reference = True
        if print_exposure is not None:
            self.enlarger.print_exposure = print_exposure
        if c_filter_neutral is not None:
            self.enlarger.c_filter_neutral = c_filter_neutral
        if m_filter_neutral is not None:
            self.enlarger.m_filter_neutral = m_filter_neutral
        if y_filter_neutral is not None:
            self.enlarger.y_filter_neutral = y_filter_neutral
        if film_density_curves is not None:
            self.film.data.density_curves = film_density_curves
            invalidates_print_balance_reference = True
        if print_density_curves is not None:
            self.print.data.density_curves = print_density_curves
        if invalidates_print_balance_reference:
            (
                self._enlarger_service.density_spectral_midgray,
                self._enlarger_service.density_spectral_midgray_comp,
            ) = self._filming_stage._compute_density_spectral_midgray_to_balance_print()
        
    # private methods
    
    def _pipeline(self, image):
        image = self._preprocess(image)
        if self.io.scan_film: # replace with route switch
            rgb_scan = self._pipeline_scan_film(image)
        else:
            rgb_scan = self._pipeline_print(image)
        return rgb_scan
    
    def _preprocess(self, image):
        image = np.double(np.array(image)[:, :, 0:3])
        image = self._filming_stage.auto_exposure(image) # autoexposure service?
        image = self._resize_service.crop_and_rescale(image)
        return image
    
    def _pipeline_scan_film(self, rgb_image):
        log_raw_film = self._filming_stage.expose(rgb_image)
        cmy_film = self._filming_stage.develop(log_raw_film)
        rgb_scan = self._scanning_stage.scan(cmy_film)
        return rgb_scan
    
    def _pipeline_print(self, rgb_image):
        log_raw_film = self._filming_stage.expose(rgb_image)
        cmy_film = self._filming_stage.develop(log_raw_film)
        log_raw_print = self._printing_stage.expose(cmy_film)
        cmy_print = self._printing_stage.develop(log_raw_print)
        rgb_scan = self._scanning_stage.scan(cmy_print)
        return rgb_scan
    
################################################################################
    
    # debug_methods

    def _pipeline_debug(self, rgb_image):
        if self.debug.debug_mode == "output":
            return self._debug_output_pipeline(rgb_image)
        elif self.debug.debug_mode == "inject":
            return self._debug_inject_pipeline(rgb_image)
    
    def _debug_output_pipeline(self, rgb_image):
        """Run the pipeline with additional outputs for debugging."""
        rgb_image = self._preprocess(rgb_image)
        log_raw_film = self._filming_stage.expose(rgb_image)
        if self.debug.output_film_log_raw:
            return log_raw_film
        
        cmy_film = self._filming_stage.develop(log_raw_film)
        if self.debug.output_film_density_cmy:
            return cmy_film
        
        log_raw_print = self._printing_stage.expose(cmy_film)
        cmy_print = self._printing_stage.develop(log_raw_print)
        if self.debug.output_print_density_cmy:
            return cmy_print
        
        rgb_scan = self._scanning_stage.scan(cmy_print)
        return rgb_scan
    
    def _debug_inject_pipeline(self, cmy_film):
        """Run the pipeline with additional inputs for debugging."""
        if self.debug.inject_film_density_cmy:
            log_raw_print = self._printing_stage.expose(cmy_film)
            cmy_print = self._printing_stage.develop(log_raw_print)
            rgb_scan = self._scanning_stage.scan(cmy_print)
            return rgb_scan

File: spectralfilm-rs/spectralfilm-core/src/runtime/pipeline.rs
//! High-level simulation pipeline orchestrator.

use std::time::Instant;

use crate::runtime::params::{DebugMode, RuntimePhotoParams};
use crate::runtime::services::{ColorReferenceService, EnlargerService, ResizingService, SpectralLUTService};
use crate::runtime::stages::{FilmingStage, PrintingStage, ScanningStage};
use crate::utils::timings::Timings;

#[derive(Debug, Clone)]
pub struct SimulationPipeline {
    params: RuntimePhotoParams,
    resize: ResizingService,
    enlarger_service: EnlargerService,
    color_reference: ColorReferenceService,
    lut_service: SpectralLUTService,
    filming: FilmingStage,
    printing: PrintingStage,
    scanning: ScanningStage,
    timings: Timings,
    last_elapsed_time: Option<f64>,
}

impl SimulationPipeline {
    pub fn new(mut params: RuntimePhotoParams) -> Self {
        params.apply_database_neutral_print_filters();
        let resize = ResizingService::new(&params.io, params.camera.film_format_mm);
        let enlarger_service = EnlargerService::new(&params.enlarger);
        let color_reference = ColorReferenceService::new(&params.scanner, &params.io);
        let lut_service = SpectralLUTService::new(params.settings.lut_resolution);
        let filming = FilmingStage::new(&params.film, &params.film_render, &params.camera, &params.io, &params.settings);
        let printing = PrintingStage::new(
            &params.film,
            &params.film_render,
            &params.print,
            &params.print_render,
            &params.enlarger,
            &params.settings,
        );
        let scanning = ScanningStage::new(
            &params.film,
            &params.film_render,
            &params.print,
            &params.print_render,
            &params.scanner,
            &params.io,
            &params.settings,
        );
        Self { params, resize, enlarger_service, color_reference, lut_service, filming, printing, scanning, timings: Timings::new(), last_elapsed_time: None }
    }

    pub fn update(&mut self, mut params: RuntimePhotoParams) {
        params.apply_database_neutral_print_filters();
        *self = Self::new(params);
    }

    pub fn process(&mut self, image: &[f64], width: usize, height: usize) -> Result<SimulationOutput, SimulationError> {
        if image.len() < width * height * 3 {
            return Err(SimulationError::InvalidInput("input buffer shorter than width*height*3".into()));
        }
        self.timings.clear();
        let total_start = Instant::now();
        
        self.color_reference.set_scan_context(self.params.io.scan_film, &self.params.film.info.r#type);
        let color_ref_clone = self.color_reference.clone();
        self.color_reference.update_positive_film_references(&self.params.film, |cmy| self.scanning.cmy_to_log_xyz(cmy, &color_ref_clone));

        let mut input = image[..width * height * 3].to_vec();
        
        let (cmy_film, w, h, pixel_size_um) = if self.params.debug.debug_mode == DebugMode::Inject && self.params.debug.inject_film_density_cmy {
            let pixel_size = self.params.camera.film_format_mm * 1000.0 / width.max(height) as f64;
            (input, width, height, pixel_size)
        } else if self.params.debug.debug_mode == DebugMode::LutGeneration {
            self.filming.auto_exposure(&mut input, width, height);
            let px_size = self.resize.pixel_size_um(); // dummy
            let log_raw_film = self.filming.expose(&input, width, height, px_size, &mut self.lut_service);
            let cmy = self.filming.develop(&log_raw_film, width, height, px_size);
            (cmy, width, height, px_size)
        } else {
            let t0 = Instant::now();
            self.filming.auto_exposure(&mut input, width, height);
            self.timings.record("FilmingStage.auto_exposure", t0.elapsed().as_secs_f64());

            let t0 = Instant::now();
            let (cropped, cw, ch) = self.resize.crop_and_rescale(&input, width, height, 3);
            self.timings.record("ResizingService.crop_and_rescale", t0.elapsed().as_secs_f64());
            let px_size = self.resize.pixel_size_um();

            let t0 = Instant::now();
            let log_raw_film = self.filming.expose(&cropped, cw, ch, px_size, &mut self.lut_service);
            self.timings.record("FilmingStage.expose", t0.elapsed().as_secs_f64());
            if self.params.debug.debug_mode == DebugMode::Output && self.params.debug.output_film_log_raw {
                self.last_elapsed_time = Some(total_start.elapsed().as_secs_f64());
                return Ok(SimulationOutput { data: log_raw_film, width: cw, height: ch, channels: 3 });
            }

            let t0 = Instant::now();
            let cmy = self.filming.develop(&log_raw_film, cw, ch, px_size);
            self.timings.record("FilmingStage.develop", t0.elapsed().as_secs_f64());
            if self.params.debug.debug_mode == DebugMode::Output && self.params.debug.output_film_density_cmy {
                self.last_elapsed_time = Some(total_start.elapsed().as_secs_f64());
                return Ok(SimulationOutput { data: cmy, width: cw, height: ch, channels: 3 });
            }
            
            (cmy, cw, ch, px_size)
        };

        let final_rgb = if self.params.io.scan_film {
            let t0 = Instant::now();
            let rgb = self.scanning.scan(&cmy_film, w, h, &mut self.lut_service, &self.color_reference);
            self.timings.record("ScanningStage.scan", t0.elapsed().as_secs_f64());
            rgb
        } else {
            let bw_filming_corr = self.color_reference.black_white_filming_exposure_correction(&self.params.film);
            let midgray_density = self.filming.compute_midgray_spectral_density(bw_filming_corr);
            let midgray_density_comp = if self.params.camera.exposure_compensation_ev != 0.0 {
                let comp = 2.0_f64.powf(self.params.camera.exposure_compensation_ev);
                Some(self.filming.compute_midgray_spectral_density(bw_filming_corr * comp))
            } else {
                None
            };
            self.enlarger_service.set_density_spectral_midgray_pair(midgray_density, midgray_density_comp);

            let t0 = Instant::now();
            let (log_raw_print, log_raw_print_black, log_raw_print_white) = self.printing.expose(
                &cmy_film, w, h, pixel_size_um, &mut self.lut_service, &self.enlarger_service, &mut self.color_reference
            );
            self.timings.record("PrintingStage.expose", t0.elapsed().as_secs_f64());

            let t0 = Instant::now();
            let cmy_print = self.printing.develop(&log_raw_print);
            self.timings.record("PrintingStage.develop", t0.elapsed().as_secs_f64());
            
            self.color_reference.set_log_raw_print_references(log_raw_print_black, log_raw_print_white);
            let color_ref_clone = self.color_reference.clone();
            self.color_reference.update_negative_print_references(&self.params.print, &self.params.print_render, |cmy| self.scanning.cmy_to_log_xyz(cmy, &color_ref_clone));

            if self.params.debug.debug_mode == DebugMode::Output && self.params.debug.output_print_density_cmy {
                self.last_elapsed_time = Some(total_start.elapsed().as_secs_f64());
                return Ok(SimulationOutput { data: cmy_print, width: w, height: h, channels: 3 });
            }

            let t0 = Instant::now();
            let rgb = self.scanning.scan(&cmy_print, w, h, &mut self.lut_service, &self.color_reference);
            self.timings.record("ScanningStage.scan", t0.elapsed().as_secs_f64());
            rgb
        };

        self.last_elapsed_time = Some(total_start.elapsed().as_secs_f64());
        Ok(SimulationOutput { data: final_rgb, width: w, height: h, channels: 3 })
    }

    pub fn get_timings(&self) -> &Timings { &self.timings }
    pub fn get_total_elapsed_time(&self) -> Option<f64> { self.last_elapsed_time }
    pub fn format_timings(&self) -> String { self.timings.format(self.last_elapsed_time) }

    /// Generates a baked 3D LUT (Look-Up Table) of the current spectral color transformation.
    /// Returns a flat vector of f32 RGB values representing the 3D lattice.
    pub fn generate_lut(&self, lut_size: usize) -> Result<Vec<f32>, SimulationError> {
        let mut lut_params = self.params.clone();
        
        // 1. Disable all spatial effects (LUTs cannot encode spatial blur)
        lut_params.film_render.grain.active = false;
        lut_params.film_render.halation.active = false;
        lut_params.film_render.glare.active = false;
        lut_params.print_render.glare.active = false;
        lut_params.camera.diffusion_filter.active = false;
        lut_params.enlarger.diffusion_filter.active = false;
        lut_params.camera.lens_blur_um = 0.0;
        lut_params.enlarger.lens_blur = 0.0;
        lut_params.scanner.lens_blur = 0.0;
        
        // 2. Disable auto-exposure (scene brightness should be handled before the LUT)
        lut_params.camera.auto_exposure = false;
        
        // 3. Set LutGeneration mode to bypass cropping
        lut_params.debug.debug_mode = DebugMode::LutGeneration;
        
        let mut pipeline = SimulationPipeline::new(lut_params);
        let n_px = lut_size * lut_size * lut_size;
        let mut grid = vec![0.0f64; n_px * 3];
        
        // 4. Generate 3D RGB lattice (R fastest, then G, then B)
        let mut idx = 0;
        for b in 0..lut_size {
            for g in 0..lut_size {
                for r in 0..lut_size {
                    grid[idx] = r as f64 / (lut_size - 1) as f64;
                    grid[idx + 1] = g as f64 / (lut_size - 1) as f64;
                    grid[idx + 2] = b as f64 / (lut_size - 1) as f64;
                    idx += 3;
                }
            }
        }
        
        // 5. Process
        let out = pipeline.process(&grid, n_px, 1)?;
        
        // 6. Convert to f32
        let f32_out: Vec<f32> = out.data.into_iter().map(|v| (v as f32).clamp(0.0, 1.0)).collect();
        Ok(f32_out)
    }
}

#[derive(Debug, Clone)]
pub struct SimulationOutput {
    pub data: Vec<f64>,
    pub width: usize,
    pub height: usize,
    pub channels: usize,
}

#[derive(Debug, Clone)]
pub enum SimulationError {
    InvalidInput(String),
}

impl std::fmt::Display for SimulationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self { Self::InvalidInput(s) => write!(f, "invalid input: {s}") }
    }
}

impl std::error::Error for SimulationError {}
